python/ir/bindings.cpp

```diff
@@ -0,0 +1,133 @@
+#include <pybind11/stl.h>
+
+#include "reffine/base/type.h"
+#include "reffine/builder/reffiner.h"
+#include "reffine/ir/expr.h"
+#include "reffine/ir/loop.h"
+#include "reffine/ir/op.h"
+#include "reffine/ir/stmt.h"
+#include "reffine/pass/printer.h"
+
+using namespace std;
+using namespace reffine;
+using namespace reffine::reffiner;
+
+namespace py = pybind11;
+
+void print_IR(shared_ptr<Func> fn)
+{
+    cout << "Loop IR:" << endl << IRPrinter::Build(fn) << endl;
+}
+
+#define REGISTER_CLASS(CLASS, PARENT, MODULE, NAME, ...)       \
+    py::class_<CLASS, shared_ptr<CLASS>, PARENT>(MODULE, NAME) \
+        .def(py::init<__VA_ARGS__>());
+
+PYBIND11_MODULE(ir, m)
+{
+    /* Structures related to Reffine typing */
+    py::enum_<BaseType>(m, "BaseType")
+        .value("bool", BaseType::BOOL)
+        .value("i8", BaseType::INT8)
+        .value("i16", BaseType::INT16)
+        .value("i32", BaseType::INT32)
+        .value("i64", BaseType::INT64)
+        .value("u8", BaseType::UINT8)
+        .value("u16", BaseType::UINT16)
+        .value("u32", BaseType::UINT32)
+        .value("u64", BaseType::UINT64)
+        .value("f32", BaseType::FLOAT32)
+        .value("f64", BaseType::FLOAT64)
+        .value("struct", BaseType::STRUCT)
+        .value("ptr", BaseType::PTR);
+
+    py::class_<DataType>(m, "DataType")
+        .def(py::init<BaseType, vector<DataType>, size_t>(), py::arg("btype"),
+             py::arg("dtypes") = vector<DataType>{}, py::arg("size") = 0)
+        .def("str", &DataType::str);
+
+    /* StmtNode and Derived Structures Declarations
+     */
+    py::class_<StmtNode, Stmt>(m, "stmt");
+    py::class_<ExprNode, Expr, StmtNode>(m, "expr");
+
+    /* Symbol Definition */
+    py::class_<SymNode, Sym, ExprNode>(m, "sym")
+        .def(py::init<string, DataType>())
+        .def(py::init<string, Expr>());
+
+    /* Constant Expressions */
+    REGISTER_CLASS(Const, ExprNode, m, "const", DataType, double)
+
+    /* Loop ops */
+    REGISTER_CLASS(IsValid, ExprNode, m, "is_valid", Expr, Expr, size_t)
+    REGISTER_CLASS(SetValid, ExprNode, m, "set_valid", Expr, Expr, Expr, size_t)
+    REGISTER_CLASS(FetchDataPtr, ExprNode, m, "fetch", Expr, Expr, size_t)
+    REGISTER_CLASS(Alloc, ExprNode, m, "alloc", DataType, Expr)
+    REGISTER_CLASS(Load, ExprNode, m, "load", Expr)
+    REGISTER_CLASS(Store, StmtNode, m, "store", Expr, Expr)
+    REGISTER_CLASS(Loop, ExprNode, m, "loop", Expr)
+
+    /* Statements */
+    REGISTER_CLASS(Func, StmtNode, m, "func", string, Expr, vector<Sym>,
+                   SymTable)
+    REGISTER_CLASS(Stmts, StmtNode, m, "stmts", vector<Stmt>)
+    REGISTER_CLASS(IfElse, StmtNode, m, "ifelse", Expr, Stmt, Stmt)
+    REGISTER_CLASS(NoOp, StmtNode, m, "noop")
+
+    /* Math Operators for Nary Expressions */
+    py::enum_<MathOp>(m, "MathOp")
+        .value("add", MathOp::ADD)
+        .value("sub", MathOp::SUB)
+        .value("mul", MathOp::MUL)
+        .value("div", MathOp::DIV)
```

Comment:
some of these keywords might be reserved in python. that is why in the CPP ir builder I defined all these with an underscore prefix.

can you please check if they work like this?

Reply:
okay. I see that these are only the enums not the actual operations. so this comment should be for the registered classes below.

Reply:
It seems to work (see [example.py](https://github.com/ampersand-projects/reffine/pull/52/files#diff-bf3cec73acca02b6a3ac711e7947bca30b2a69256664162042e680489af89e61)) but I can add the underscore just to make sure it doesn't cause any trouble in the future.

Reply:
I added to the enum as well for consistency.